x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStopTrainedModelDeploymentAction.java

```diff
@@ -117,13 +121,39 @@ private void redirectToMasterNode(DiscoveryNode masterNode, StopTrainedModelDepl
         }
     }
 
-    private void normalUndeploy(Task task, PersistentTasksCustomMetadata.PersistentTask<?> deployTrainedModelTask,
+    private void normalUndeploy(Task task,
+                                String modelId,
+                                TrainedModelAllocation modelAllocation,
                                 StopTrainedModelDeploymentAction.Request request,
                                 ActionListener<StopTrainedModelDeploymentAction.Response> listener) {
-        request.setNodes(deployTrainedModelTask.getExecutorNode());
-
+        request.setNodes(modelAllocation.getNodeRoutingTable().keySet().toArray(String[]::new));
         ActionListener<StopTrainedModelDeploymentAction.Response> finalListener = ActionListener.wrap(
-            r -> waitForTaskRemoved(Collections.singleton(deployTrainedModelTask.getId()), request, r, listener),
+            r -> {
+                waitForTaskRemoved(modelId, modelAllocation, request, r, ActionListener.wrap(
+                    waited -> {
+                        trainedModelAllocationService.deleteModelAllocation(
+                            modelId,
+                            ActionListener.wrap(
+                                deleted -> listener.onResponse(r),
+                                deletionFailed -> {
+                                    logger.error(
+                                        () -> new ParameterizedMessage(
+                                            "[{}] failed to delete model allocation after nodes unallocated the deployment",
+                                            modelId
+                                        ),deletionFailed);
+                                    listener.onFailure(ExceptionsHelper.serverError(
+                                        "failed to delete model allocation after nodes unallocated the deployment. Attempt to stop again",
+                                        deletionFailed
+                                    ));
+                                }
+                            )
+                        );
+                    },
+                    // TODO should we attempt to delete the deployment here?
+                    listener::onFailure
```

Comment:
If we failed to stop on nodes, we probably should still delete the allocation as that will cause the individual nodes to also react and cancel tasks. It just may end up in a weird state for a small bit.